### PR TITLE
Updates Scala specification for 2.10

### PR DESCRIPTION
**Pull request by soc**
- Removed Octal literals
- Disallowed FP literals without digit after dot
- Removed val in for comprehension value definitions (Fixes SI-4918)
- Added ## to the list of non-NPE-throwing Null methods (Fixes SI-4311)
- Adapted some wording around AnyRef/AnyVal
- Removed primitive type aliases (Fixes SI-4628)
- Casting of null (Fixes second part of SI-4437)
- Removed ScalaObject from the spec (Closes SI-4648)
- Merged the changes of the PDF-only Scala-2.9-Draft manually

### CODE REVIEW

**Comment by jsuereth** [comment]
Review by: @odersky

**Comment by soc** [comment]
Any news on this?

**Comment by soc** [comment]
- Removed Octal literals is SI-5205
- Disallowed FP literals without digit after dot is related to SI-265 (found no better issue)

I added links from the issues to this pull request, so that people don't mistakenly duplicate work.

File: documentation/src/reference/ReferencePart.tex [hunk 1]
```diff
@@ -5777,25 +5757,7 @@ \subsection{Eta Expansion}\label{sec:eta-expand}
...
-Further assuming the selection is not followed by any function arguments, such an expression can be rewitten under the conditions given in \sref{sec:impl-conv} to:
-\begin{lstlisting}
-$e$.applyDynamic("$x$")
-\end{lstlisting}
-If the selection is followed by some arguments, e.g.\ $e.x(\args)$, then that expression
-is rewritten to
-\begin{lstlisting}
-$e$.applyDynamic("$x$", $\args$)
-\end{lstlisting}
+% TODO
```
Comment (Blaisorblade) [inline documentation/src/reference/ReferencePart.tex:5760]: This doesn't seem mentioned in the issue description, and Scala does have `scala.Dynamic` (with some slightly different details). So I'm against merging this bit.

**Comment by Blaisorblade** [comment]
I'd hope that some technical uncontroversial changes (say, to the grammar) could be merged after review by @adriaanm. If that were possible (and I think that should be possible and overdue), probably splitting somewhat the changes of this commit would help reviewing. Personally I'd have a commit per fixed issue, but that's my taste.

OTOH, I can totally get why @odersky would restrict maintenance of, say, the type system spec to type theory experts (or even to himself).

@soc wrote:

> Merged the changes of the PDF-only Scala-2.9-Draft manually

Is that why you removed so much from `documentation/src/reference/ImplementationStatus.tex` and `documentation/src/reference/RationalePart.tex`? Because such changes can only be reviewed by @odersky himself, unless he already did them.

So I'd say that should be a separate commit at least for this part, also to ease reviewing.

Also, why is so much removed from "Dynamic Member Selection"?

**Comment by Blaisorblade** [comment]
> OTOH, I can totally get why @odersky would restrict maintenance of, say, the type system spec to type theory experts (or even to himself).

But yet again OTOH, I'd say that changes like https://github.com/scala/scala/commit/e28c3edda4 should also be reflected in the spec: I'd like to know what exactly that change does, and I wouldn't want to debug a type error following the affected part of the spec.

**Comment by soc** [comment]
@Blaisorblade I don't think the spec is updated anymore in its current form. I guess this commit probably only exists to remind people of the sad state. Sadly, I have heard no news about the new spec format since the guy who did it was hired (by Google?!) ... not sure about the details anymore.

**Comment by Blaisorblade** [comment]
Last I knew, Martin had exclusive maintainership of the spec and was swamped on other work, I thought that was the reason (can't find the message in scala-internals or Google unfortunately). And a (partial? abandoned?) migration to some new format sounds a less valid argument. Would it be possible to do something?

**Comment by soc** [comment]
Well, there is not much I can do about it ... I have accepted that the turn-around time has to be measured in months (spec), weeks (code) or years (website).

**Comment by adriaanm** [comment]
I have scheduled some time to update the spec, but due to limited resources it'll have to wait until M5 at the earliest.

**Comment by soc** [comment]
Nice to hear! (Although this pull request is probably not what you are looking for.) I had already split up the changes into 1 PR per change, but there hasn't been much progress recently (Iain seems to be pretty busy + more reasons here: https://github.com/iainmcgin/scala-ref-markdown/pull/19): https://github.com/iainmcgin/scala-ref-markdown/pulls

**Comment by soc** [comment]
@odersky Done. See #97, #98, #99, #100, #101, #102, #103.

**Pull request closed without merging**